src/ir/possible-contents.h

```diff
@@ -0,0 +1,577 @@
+        value = ExactType(Type(otherType.getHeapType(), Nullable));
+        return;
+      } else if (isNull() && other.isNull()) {
+        // Both are null. The result is a null, of the LUB.
+        auto lub = HeapType::getLeastUpperBound(type.getHeapType(),
+                                                otherType.getHeapType());
+        value = Literal::makeNull(lub);
+        return;
+      }
+    }
+
+    if (isTypeExact() && other.isTypeExact() &&
+        type.getHeapType() == otherType.getHeapType()) {
+      // We know the types here exactly, and even the heap types match, but
+      // there is some other difference that prevents them from being 100%
+      // identical (for example, one might be an ExactType and the other a
+      // Literal; or both might be ExactTypes and only one might be nullable).
+      // In these cases we can emit a proper ExactType here, adding nullability
+      // if we need to.
+      value = ExactType(Type(
+        type.getHeapType(),
+        type.isNullable() || otherType.isNullable() ? Nullable : NonNullable));
+      return;
+    }
+
+    // Nothing else possible combines in an interesting way; emit a Many.
+    value = Many();
+  }
+
+  bool isNone() const { return std::get_if<None>(&value); }
+  bool isLiteral() const { return std::get_if<Literal>(&value); }
+  bool isGlobal() const { return std::get_if<GlobalInfo>(&value); }
+  bool isExactType() const { return std::get_if<Type>(&value); }
+  bool isMany() const { return std::get_if<Many>(&value); }
+
+  Literal getLiteral() const {
+    assert(isLiteral());
+    return std::get<Literal>(value);
+  }
+
+  Name getGlobal() const {
+    assert(isGlobal());
+    return std::get<GlobalInfo>(value).name;
+  }
+
+  bool isNull() const { return isLiteral() && getLiteral().isNull(); }
+
+  // Return the relevant type here. Note that the *meaning* of the type varies
+  // by the contents: type $foo of a global means that type or any subtype, as a
+  // subtype might be written to it, while type $foo of a Literal or an
+  // ExactType means that type and nothing else; see isTypeExact().
+  //
+  // If no type is possible, return unreachable; if many types are, return none.
+  Type getType() const {
+    if (auto* literal = std::get_if<Literal>(&value)) {
+      return literal->type;
+    } else if (auto* global = std::get_if<GlobalInfo>(&value)) {
+      return global->type;
+    } else if (auto* type = std::get_if<Type>(&value)) {
+      return *type;
+    } else if (std::get_if<None>(&value)) {
+      return Type::unreachable;
+    } else if (std::get_if<Many>(&value)) {
+      return Type::none;
+    } else {
+      WASM_UNREACHABLE("bad value");
+    }
+  }
+
+  // Returns whether the type we can report here is exact, that is, nothing of a
+  // strict subtype might show up.
+  //
+  // This returns false for None and Many, for whom it is not well-defined.
+  bool isTypeExact() const { return isExactType() || isLiteral(); }
```

Comment:
This is confusingly similar to `isExactType`. Can we change it to `hasExactType` maybe? Or maybe this should be rolled into `getExactType` as suggested above so that it covers both `ExactType` and `Literal`.

Reply:
I agree the current name has some risk of confusion. `hasExactType` can also make sense, interesting idea. I feel like it has similar risks, though, so I don't have a preference.

@aheejin `isExactType` checks if something is an exact type, and then we have two options for a function that checks if something's type is known exactly (which is true for ExactType, but also constant values as well):
* `isTypeExact` (sort of "is the type exact")
* `hasExactType` (sort of "does it have an exact type")

Which do you think is clearer?

Reply:
`isTypeExact` wasn't very confusing when I read it, so I don't have a strong preference, but I guess I slightly prefer the latter.

Reply:
Sounds good, thanks, I'll change it to `hasExactType` then.